packages/grafana-prometheus/src/components/monaco-query-field/monaco-completion-provider/completions.ts

```diff
@@ -16,25 +22,45 @@ type Completion = {
   triggerOnInsert?: boolean;
 };
 
-type Metric = {
-  name: string;
-  help: string;
-  type: string;
-};
+const metricNamesSearchClient = new UFuzzy({
+  intraMode: 1,
+  intraChars: '[a-z0-9_:]', // characters allowed in metric names, according to the Prometheus data model
+});
 
-export type DataProvider = {
-  getHistory: () => Promise<string[]>;
-  getAllMetricNames: () => Promise<Metric[]>;
-  getAllLabelNames: () => Promise<string[]>;
-  getLabelValues: (labelName: string) => Promise<string[]>;
-  getSeriesValues: (name: string, match: string) => Promise<string[]>;
-  getSeriesLabels: (selector: string, otherLabels: Label[]) => Promise<string[]>;
-};
+export function fuzzySearchMetrics(haystack: Metric[], needle: string): Metric[] {
+  const filteredMetrics: Metric[] = [];
+  const metricNames = haystack.map((metric) => metric.name);
+  const [idxs] = metricNamesSearchClient.search(metricNames, needle);
 
-// we order items like: history, functions, metrics
+  for (const idx of idxs ?? []) {
+    const metric = haystack.at(idx);
```

Comment:
For curious minds: array access is ~11% faster than `Array.prototype.at`:

![jsperf-array-access-vs-array-prototype-at](https://github.com/grafana/grafana/assets/5732000/e303c928-cb5a-4190-ba98-344a35317c94)

Reply:
what does this look like with 10k values and with a `for` loop instead of forEach callback?

Reply:
thanks!

good to know. last time i checked .at() was like 50% slower. maybe it's catching up :)

generally when iterating possibly thousands of values, i try to stick with for loops and direct access since they're the most optimized in js VMs.